File: tools/core070_postfit_policy_batch.R
# Retained evidence runner for the "postfit-policy" manifest-area batch (15
# EXECUTABLE_NOW cases + 2 negative controls; see
# docs/dev-log/core070/postfit-policy-batch-contract.json for the full case
# list, the 8 NEEDS_NEW_JULIA_SURFACE deferrals, and the 1 zero-case
# SPEC_DEFECT row).
#
# REPAIR NOTE (2026-09-01, Totoro incident): the prior version of this pair
# had the Julia child (tools/core070_postfit_policy_batch.jl) `include()`
# the ENTIRE test/parity/runparity.jl runner and embed a live R session via
# RCall to obtain the R oracle numbers. On Totoro that first segfaulted
# (exit 139 -- cured only by LD_PRELOAD=<julia>/lib/julia/libunwind.so.8 for
# the *subprocess* julia, an unrelated libunwind/RCall interaction worth
# keeping as a comment here in case a future runner hits it again), then
# failed inside test/parity/test_negbin_parity.jl -- an out-of-scope RCall
# fixture nothing in this 15-case policy batch needs. Rewritten so this R
# process (the one that ALREADY has the frozen gllvmTMB library loaded) does
# 100% of the R-side computation itself and hands the Julia child a plain
# JSON file (Y + oracle numbers); the Julia child now runs zero R code and
# carries no RCall dependency at all -- see its own header for the read side.
#
# argv (matches the contract's runner.outer_argv and the --self-test local
# smoke, which never reaches this file):
#   Rscript --vanilla tools/core070_postfit_policy_batch.R <frozen-library> <destination>
#
# <frozen-library> is an R library directory containing an installed
# gllvmTMB built from the pinned reference commit (b4d5fee...); this is the
# FROZEN, INSTALLED library, matching masks_known.R's arg 1 -- not a source
# tree. <destination> must not already exist; it is created and holds
# r-oracle.json, julia-results.json, julia-stdout.log, julia-stderr.log,
# results.tsv, diagnostics.log, and receipt.json.
#
# NOTE on the contract's documented `inner_argv`/`inner_invocation` strings:
# those describe the PRE-repair design (Julia takes one positional arg, and
# reads R state via RCall/env-var-gated runparity.jl inclusion). Positional
# argv to the Julia child is unchanged (still exactly one path, the results
# destination); the new R-oracle JSON travels via the
# CORE070_POSTFIT_POLICY_R_ORACLE env var instead of RCall, so the
# documented single-positional-arg shape still holds. Recorded here, not
# smoothed over, since the contract JSON itself is frozen and not edited by
# this repair.

args <- commandArgs(TRUE)
stopifnot(length(args) == 2L)
frozen_library <- normalizePath(args[[1]], mustWork = TRUE)
output_dir <- args[[2]]
stopifnot(!dir.exists(output_dir))
dir.create(output_dir, recursive = TRUE)

sha256_file <- function(path) {
  command <- if (nzchar(Sys.which("sha256sum"))) "sha256sum" else "shasum"
  argv <- if (identical(command, "sha256sum")) path else c("-a", "256", path)
  line <- system2(command, argv, stdout = TRUE, stderr = TRUE)
  stopifnot(is.null(attr(line, "status")), length(line) >= 1L)
  sub("[[:space:]].*$", "", line[[1L]])
}

.libPaths(c(frozen_library, .libPaths()))
suppressPackageStartupMessages(library(gllvmTMB))
suppressPackageStartupMessages(library(jsonlite))
stopifnot(normalizePath(find.package("gllvmTMB")) ==
          normalizePath(file.path(frozen_library, "gllvmTMB")))

root <- normalizePath(".")
contract_path <- file.path(root, "docs/dev-log/core070/postfit-policy-batch-contract.json")
contract <- jsonlite::read_json(contract_path, simplifyVector = FALSE)
contract_sha256 <- sha256_file(contract_path)

stopifnot(identical(contract$reference_commit, "b4d5fee64def88bc768dda1f1f77c29b295edd86"),
          identical(contract$status, "FROZEN_POSTFIT_POLICY_BATCH_CONTRACT"),
          length(contract$cases) == contract$expected_case_count,
          contract$expected_case_count == 15L,
          length(contract$negative_controls) >= 2L)

# --- validate pinned R source ------------------------------------------
source_root <- file.path(root, ".unlazy/core070-aghq/oracle-source/readback")
for (rel in names(contract$source_pins)) {
  path <- file.path(source_root, rel)
  stopifnot(file.exists(path))
  digest <- sha256_file(path)
  stopifnot(identical(digest, contract$source_pins[[rel]]))
}

# ---------------------------------------------------------------------------
# 1. Fixture: p x n Gaussian data, same design formula the rest of the
#    core070 postfit tools use (value ~ 0 + trait + latent(0 + trait | site,
#    d = K, unique = FALSE)); matches the contract fixture note's p_times_n
#    = 400 (p = 5, K = 2, n = 80). This IS the R side's own Y -- the Julia
#    child reads it back verbatim from the oracle JSON below and refits it
#    natively, so no cross-language RNG matching is required anywhere.
# ---------------------------------------------------------------------------
set.seed(42)
p <- 5L; K <- 2L; n <- 80L
Lambda_true <- matrix(c(
  0.8,  0.0,
  0.5,  0.6,
  0.3, -0.4,
 -0.2,  0.5,
  0.1,  0.3
), nrow = p, ncol = K, byrow = TRUE)
sigma_true <- 0.7
eta <- matrix(rnorm(K * n), nrow = K, ncol = n)
Y <- Lambda_true %*% eta + sigma_true * matrix(rnorm(p * n), nrow = p, ncol = n)
Y <- Y - rowMeans(Y)

trait_names <- paste0("t", seq_len(p))
df_long <- data.frame(
  site  = factor(rep(seq_len(n), each = p)),
  trait = factor(rep(trait_names, times = n), levels = trait_names),
  value = as.vector(Y)
)

fit_r <- gllvmTMB(
  value ~ 0 + trait + latent(0 + trait | site, d = K, unique = FALSE),
  data = df_long, unit = "site", trait = "trait", family = gaussian(),
  control = gllvmTMBcontrol(n_init = 1L, se = TRUE)
)
stopifnot("gllvmTMB_multi" %in% class(fit_r))

# ---------------------------------------------------------------------------
# 2. R-side accessor + reflection oracle (the exact set the 15 executable
#    cases need; ported unchanged from the pre-repair Julia-embedded R block).
# ---------------------------------------------------------------------------
.pp_get3 <- function(generic, cls) {
  m <- tryCatch(utils::getS3method(generic, cls, envir = asNamespace("gllvmTMB")),
                error = function(e) NULL)
  if (is.null(m)) m <- get(paste0(generic, ".", cls), envir = asNamespace("gllvmTMB"))
  m
}
.pp_predict_fun   <- .pp_get3("predict", "gllvmTMB_multi")
.pp_residuals_fun <- .pp_get3("residuals", "gllvmTMB_multi")
.pp_simulate_fun  <- .pp_get3("simulate", "gllvmTMB_multi")

oracle <- list(
  coef      = as.numeric(coef(fit_r)),
  nobs      = as.integer(nobs(fit_r)),
  df        = as.integer(attr(logLik(fit_r), "df")),
  loglik    = as.numeric(logLik(fit_r)),
  loglik_nobs_attr = as.integer(attr(logLik(fit_r), "nobs")),
  link      = as.numeric(predict(fit_r, type = "link")$est),
  response  = as.numeric(fitted(fit_r)$est),
  residual  = as.numeric(residuals(fit_r, type = "randomized_quantile", scale = "normal")$residual),
  predict_type_default   = as.character(eval(formals(.pp_predict_fun)$type)[1]),
  residual_type_default  = as.character(eval(formals(.pp_residuals_fun)$type)[1]),
  residual_scale_default = as.character(eval(formals(.pp_residuals_fun)$scale)[1]),
  simulate_condition_on_re_default = isFALSE(formals(.pp_simulate_fun)$condition_on_RE)
)

ci <- tryCatch({
  cc <- confint(fit_r, parm = fit_r$X_fix_names, method = "wald")
  list(ok = TRUE, lower = as.numeric(cc[, 1]), upper = as.numeric(cc[, 2]), error = "")
}, error = function(e) list(ok = FALSE, lower = numeric(0), upper = numeric(0),
                             error = conditionMessage(e)))

# --- POST-COEF-EMPTY: parse-one-function-and-eval on a synthetic mock ------
.pp_coef_defs <- Filter(
  function(x) is.call(x) && identical(x[[1L]], as.name("<-")) &&
    identical(x[[2L]], as.name("coef.gllvmTMB_multi")),
  parse(file.path(source_root, "R/vcov-coef.R"))
)
stopifnot(length(.pp_coef_defs) == 1L)
.pp_coef_env <- new.env(parent = asNamespace("gllvmTMB"))
eval(.pp_coef_defs[[1L]], .pp_coef_env)
empty_coef <- .pp_coef_env$coef.gllvmTMB_multi(list(X_fix_names = character(0)))

# ---------------------------------------------------------------------------
# 3. Write the oracle JSON (Y + all R-side values) for the Julia child.
# ---------------------------------------------------------------------------
oracle_path <- file.path(output_dir, "r-oracle.json")
jsonlite::write_json(
  list(
    schema = "core070-postfit-policy-r-oracle/v1",
    p = p, n = n, K = K,
    y = as.numeric(Y),               # column-major flatten; Julia reshapes (p, n)
    coef = oracle$coef,
    nobs = oracle$nobs,
    df = oracle$df,
    loglik = oracle$loglik,
    loglik_nobs_attr = oracle$loglik_nobs_attr,
    link = oracle$link,              # column-major flatten of the (p, n) matrix
    response = oracle$response,
    residual = oracle$residual,
    predict_type_default = oracle$predict_type_default,
    residual_type_default = oracle$residual_type_default,
    residual_scale_default = oracle$residual_scale_default,
    simulate_condition_on_re_default = oracle$simulate_condition_on_re_default,
    ci_ok = ci$ok, ci_lower = ci$lower, ci_upper = ci$upper, ci_error = ci$error,
    empty_coef = as.numeric(empty_coef)
  ),
  oracle_path, auto_unbox = TRUE, pretty = TRUE, null = "null", na = "null", digits = 17
)

# --- invoke the Julia child ---------------------------------------------
julia_out <- file.path(output_dir, "julia-results.json")
julia_env <- c(CORE070_POSTFIT_POLICY_R_ORACLE = oracle_path)
julia_bin <- Sys.which("julia")
stopifnot(nzchar(julia_bin))
old_env <- Sys.getenv(names(julia_env), unset = NA, names = TRUE)
do.call(Sys.setenv, as.list(julia_env))
t0 <- Sys.time()
julia_status <- tryCatch(
  system2(julia_bin, c("--project=.", "tools/core070_postfit_policy_batch.jl", julia_out),
          stdout = file.path(output_dir, "julia-stdout.log"),
          stderr = file.path(output_dir, "julia-stderr.log")),
  finally = {
    unset <- names(old_env)[is.na(old_env)]
    if (length(unset)) Sys.unsetenv(unset)
    set <- old_env[!is.na(old_env)]
    if (length(set)) do.call(Sys.setenv, as.list(set))
  }
)
elapsed <- as.numeric(Sys.time() - t0, units = "secs")

# --- read back and evaluate ----------------------------------------------
julia_report <- if (file.exists(julia_out)) {
  jsonlite::read_json(julia_out, simplifyVector = FALSE)
} else {
  NULL
}

contract_case_ids <- vapply(contract$cases, `[[`, "", "case_id")
contract_neg_ids <- vapply(contract$negative_controls, `[[`, "", "control_id")

results_ok <- !is.null(julia_report) &&
  identical(julia_status, 0L) &&
  identical(julia_report$status, "PASS") &&
  isTRUE(julia_report$all_positive_pass) &&
  isTRUE(julia_report$negative_controls_behaved_as_expected) &&
  isTRUE(julia_report$all_checks) &&
  identical(sort(names(julia_report$cases)), sort(contract_case_ids)) &&
  identical(sort(names(julia_report$negative_controls)), sort(contract_neg_ids)) &&
  all(vapply(julia_report$cases, function(x) isTRUE(x$pass), logical(1))) &&
  all(vapply(julia_report$negative_controls, function(x) isTRUE(x$behaved), logical(1)))

raw_lines <- if (!is.null(julia_report)) {
  vapply(contract_case_ids, function(id) {
    pass <- isTRUE(julia_report$cases[[id]]$pass)
    paste(id, if (pass) "PASS" else "FAIL", "positive", sep = "\t")
  }, character(1))
} else character(0)
neg_lines <- if (!is.null(julia_report)) {
  vapply(contract_neg_ids, function(id) {
    behaved <- isTRUE(julia_report$negative_controls[[id]]$behaved)
    paste(id, if (behaved) "PASS" else "FAIL", "negative_control", sep = "\t")
  }, character(1))
} else character(0)
raw_path <- file.path(output_dir, "results.tsv")
writeLines(c(raw_lines, neg_lines), raw_path)

diag_lines <- character(0)
if (!identical(julia_status, 0L)) diag_lines <- c(diag_lines, paste("julia_exit_code", julia_status))
if (is.null(julia_report)) diag_lines <- c(diag_lines, "julia-results.json was not written")
diag_path <- file.path(output_dir, "diagnostics.log")
writeLines(diag_lines, diag_path)

receipt <- list(
  status = if (isTRUE(results_ok)) "PASS" else "FAIL",
  scope = "CORE070_POSTFIT_POLICY_BATCH",
  reference_commit = contract$reference_commit,
  contract_sha256 = contract_sha256,
  source_pins = contract$source_pins,
  source_unchanged = TRUE,
  case_count = contract$expected_case_count,
  negative_control_count = length(contract$negative_controls),
  expected_case_ids = contract_case_ids,
  negative_control_case_ids = contract_neg_ids,
  julia_exit_code = julia_status,
  julia_elapsed_seconds = elapsed,
  julia_results_sha256 = if (file.exists(julia_out)) sha256_file(julia_out) else NA_character_,
  raw_sha256 = sha256_file(raw_path),
  diagnostics_sha256 = sha256_file(diag_path),
  r_version = R.version.string,
  gllvmTMB_version = as.character(utils::packageVersion("gllvmTMB")),
  frozen_library = frozen_library
)
receipt_path <- file.path(output_dir, "receipt.json")
jsonlite::write_json(receipt, receipt_path, auto_unbox = TRUE, pretty = TRUE, null = "null")

cat("CORE070_POSTFIT_POLICY_BATCH_", receipt$status, "\n", sep = "")
quit(status = if (isTRUE(results_ok)) 0L else 1L)

File: tools/core070_postfit_policy_batch.jl
# Inner executor for the "postfit-policy" manifest-area batch (15 planned
# EXECUTABLE_NOW cases + 2 negative controls).
#
# REPAIR NOTE (2026-09-01, Totoro incident): the prior version of this file
# `include()`d the ENTIRE test/parity/runparity.jl parity runner and used
# RCall to embed a live R session, obtaining the R oracle numbers by @rput
# fixture data into it and driving gllvmTMB(). On Totoro that first
# segfaulted (exit 139 for the subprocess julia process; cured only by
# LD_PRELOAD=<julia>/lib/julia/libunwind.so.8 -- an RCall/libunwind
# interaction worth remembering if a future runner hits the same signal),
# then failed inside test/parity/test_negbin_parity.jl -- an out-of-scope
# RCall fixture the runparity.jl runner reaches internally, well beyond this
# batch's 15-case scope. This file is now a pure-Julia consumer: it reads a
# JSON oracle file the paired R runner (tools/core070_postfit_policy_batch.R)
# writes BEFORE invoking this script -- that R process already has the
# frozen gllvmTMB library loaded and does 100% of the live R fitting itself
# -- refits the identical Y natively with fit_gaussian_gllvm, and compares
# via direct `using GLLVM` module calls only. No RCall, no parity-runner
# include, no R of any kind runs in this process.
#
# argv:
#   ARGS[1]  destination path for the JSON results file (parent dir need not
#            exist yet; this script mkpath()s it). Must not already exist.
#
# Env vars required (set by the outer R runner before invoking this script;
# never defaulted silently here so a misconfigured environment fails loudly
# rather than quietly running against stale or missing oracle data):
#   CORE070_POSTFIT_POLICY_R_ORACLE = <path to the R-written oracle JSON>
#
# Invocation:
#   julia --project=. tools/core070_postfit_policy_batch.jl <out.json>

using GLLVM

# ---------------------------------------------------------------------------
# Minimal JSON reader/writer (no external dependency; mirrors the existing
# repo convention in tools/core070_postfit_1_batch.jl / core070_data_batch.jl).
# ---------------------------------------------------------------------------
function json_read(path::AbstractString)
    txt = read(path, String)
    pos = Ref(1)
    skip_ws!() = begin
        while pos[] <= lastindex(txt) && isspace(txt[pos[]])
            pos[] = nextind(txt, pos[])
        end
    end
    local parse_value
    function parse_string()
        pos[] += 1 # opening quote
        buf = IOBuffer()
        while true
            c = txt[pos[]]
            if c == '"'
                pos[] = nextind(txt, pos[])
                break
            elseif c == '\\'
                pos[] = nextind(txt, pos[])
                e = txt[pos[]]
                if e == 'n'; write(buf, '\n')
                elseif e == 't'; write(buf, '\t')
                elseif e == 'u'
                    hex = txt[nextind(txt, pos[]):nextind(txt, pos[], 4)]
                    write(buf, Char(parse(UInt32, hex; base = 16)))
                    pos[] = nextind(txt, pos[], 4)
                else write(buf, e)
                end
                pos[] = nextind(txt, pos[])
            else
                write(buf, c)
                pos[] = nextind(txt, pos[])
            end
        end
        String(take!(buf))
    end
    function parse_number()
        start = pos[]
        while pos[] <= lastindex(txt) && (isdigit(txt[pos[]]) || txt[pos[]] in ('-', '+', '.', 'e', 'E'))
            pos[] = nextind(txt, pos[])
        end
        s = txt[start:prevind(txt, pos[])]
        return occursin('.', s) || occursin('e', s) || occursin('E', s) ? parse(Float64, s) : parse(Int, s)
    end
    function parse_array()
        pos[] += 1
        out = Any[]
        skip_ws!()
        if txt[pos[]] == ']'
            pos[] = nextind(txt, pos[])
            return out
        end
        while true
            skip_ws!()
            push!(out, parse_value())
            skip_ws!()
            if txt[pos[]] == ','
                pos[] = nextind(txt, pos[])
            elseif txt[pos[]] == ']'
                pos[] = nextind(txt, pos[])
                break
            end
        end
        out
    end
    function parse_object()
        pos[] += 1
        out = Dict{String, Any}()
        skip_ws!()
        if txt[pos[]] == '}'
            pos[] = nextind(txt, pos[])
            return out
        end
        while true
            skip_ws!()
            k = parse_string()
            skip_ws!()
            @assert txt[pos[]] == ':'
            pos[] = nextind(txt, pos[])
            skip_ws!()
            out[k] = parse_value()
            skip_ws!()
            if txt[pos[]] == ','
                pos[] = nextind(txt, pos[])
            elseif txt[pos[]] == '}'
                pos[] = nextind(txt, pos[])
                break
            end
        end
        out
    end
    parse_value = function ()
        skip_ws!()
        c = txt[pos[]]
        if c == '"'; return parse_string()
        elseif c == '['; return parse_array()
        elseif c == '{'; return parse_object()
        elseif c == 't'; pos[] += 4; return true
        elseif c == 'f'; pos[] += 5; return false
        elseif c == 'n'; pos[] += 4; return nothing
        else; return parse_number()
        end
    end
    skip_ws!()
    parse_value()
end

json_escape(s::AbstractString) = replace(s, "\\" => "\\\\", "\"" => "\\\"", "\n" => "\\n")
to_json(x::Bool) = x ? "true" : "false"
to_json(x::Nothing) = "null"
to_json(x::AbstractString) = "\"" * json_escape(x) * "\""
to_json(x::Integer) = string(x)
to_json(x::AbstractFloat) = isfinite(x) ? repr(x) : "null"
to_json(x::AbstractVector) = "[" * join(to_json.(x), ",") * "]"
to_json(x::AbstractDict) = "{" * join(("\"$(json_escape(string(k)))\":" * to_json(v) for (k, v) in x), ",") * "}"

length(ARGS) == 1 || error("usage: julia tools/core070_postfit_policy_batch.jl <out.json>")
out_path = ARGS[1]
isfile(out_path) && error("destination already exists: $out_path")
mkpath(dirname(out_path))

@assert realpath(Base.pkgdir(GLLVM)) == realpath(pwd()) "must run from the GLLVM.jl package root"

oracle_path = get(ENV, "CORE070_POSTFIT_POLICY_R_ORACLE", "")
isempty(oracle_path) &&
    error("CORE070_POSTFIT_POLICY_R_ORACLE is required; refusing to silently default it")
isfile(oracle_path) ||
    error("CORE070_POSTFIT_POLICY_R_ORACLE points to a nonexistent file: $oracle_path")

oracle = json_read(oracle_path)

# ---------------------------------------------------------------------------
# Read the R-written fixture and oracle numbers. p/n/K/Y come straight from
# the R side's own live fit -- no independent regeneration, no cross-language
# RNG matching required.
# ---------------------------------------------------------------------------
p = Int(oracle["p"])
n = Int(oracle["n"])
K = Int(oracle["K"])
y_flat = Float64.(oracle["y"])
length(y_flat) == p * n || error("oracle y length ($(length(y_flat))) != p*n ($(p*n))")
Y = reshape(y_flat, p, n) # R wrote column-major (p, n): matches Julia reshape directly

X = zeros(p, n, p)
for j in 1:p
    X[j, :, j] .= 1
end

r_coef = Float64.(oracle["coef"])
r_nobs = Int(oracle["nobs"])
r_df = Int(oracle["df"])
r_loglik = Float64(oracle["loglik"])
r_loglik_nobs_attr = Int(oracle["loglik_nobs_attr"])
r_link = reshape(Float64.(oracle["link"]), size(Y))
r_response = reshape(Float64.(oracle["response"]), size(Y))
r_residual = reshape(Float64.(oracle["residual"]), size(Y))
r_predict_type_default = String(oracle["predict_type_default"])
r_residual_type_default = String(oracle["residual_type_default"])
r_residual_scale_default = String(oracle["residual_scale_default"])
r_simulate_condition_on_re_default = Bool(oracle["simulate_condition_on_re_default"])
r_ci_ok = Bool(oracle["ci_ok"])
r_ci_lower = Float64.(oracle["ci_lower"])
r_ci_upper = Float64.(oracle["ci_upper"])
r_ci_error = String(oracle["ci_error"])
r_empty_coef = Float64.(oracle["empty_coef"])

# ---------------------------------------------------------------------------
# Native Julia refit + direct GLLVM module accessor calls (no RCall).
# ---------------------------------------------------------------------------
fit = fit_gaussian_gllvm(Y; K = K, X = X)

j_coef = GLLVM.StatsAPI.coef(fit)
j_loglik = fit.logLik
j_dof = GLLVM.StatsAPI.dof(fit)
j_nobs = GLLVM.StatsAPI.nobs(fit, Y)
j_link = predict(fit, Y; type = :link, X = X)
j_response = GLLVM.StatsAPI.fitted(fit, Y; X = X)
j_residual = residuals(fit, Y; X = X)
j_ci = try
    GLLVM.confint(fit; y = Y, X = X)
catch e
    nothing
end
j_ci_ok = j_ci !== nothing
beta_idx = j_ci_ok ? findall(t -> startswith(t, "beta["), j_ci.term) : Int[]
j_ci_lower = j_ci_ok ? j_ci.lower[beta_idx] : Float64[]
j_ci_upper = j_ci_ok ? j_ci.upper[beta_idx] : Float64[]

# simulate() keyword introspection: does GLLVM's simulate accept a
# condition_on_RE-style keyword at all? (it must not, per the manifest fact)
# Call-based probe (not Base.kwarg_decl, which is an unstable internal): a
# MethodError from an unrecognized keyword means the kwarg is absent; any
# other outcome (success, or an error raised from inside a method that DID
# accept the keyword) means it exists.
julia_simulate_has_condition_on_re_kwarg = try
    GLLVM.simulate(fit, 2; condition_on_RE = true)
    true
catch err
    !(err isa MethodError)
end

# ---------------------------------------------------------------------------
# Empty-design pair (POST-COEF-EMPTY): a real no-X Julia fit vs the pinned
# R coef.gllvmTMB_multi empty branch (already evaluated by the R stage
# against a synthetic mock object; r_empty_coef comes from the oracle file).
# ---------------------------------------------------------------------------
fit_no_x = fit_gaussian_gllvm(Y; K = 1)
j_empty_coef = GLLVM.StatsAPI.coef(fit_no_x)

# ---------------------------------------------------------------------------
# Assemble the 15 case results.
# ---------------------------------------------------------------------------
tol = Dict(
    "coefficient_delta" => 1e-6,
    "loglik_delta" => 1e-6,
    "link_response_residual_delta" => 1e-4,
    "wald_ci_bound_delta" => 1e-3,
)

cases = Dict{String, Any}()

cases["CORE070-POSTFIT-COEF-EMPTY-NATIVE"] = Dict(
    "pass" => length(j_empty_coef) == 0 && length(r_empty_coef) == 0,
    "julia_length" => length(j_empty_coef), "r_length" => length(r_empty_coef),
)

coefficient_delta = maximum(abs.(j_coef .- r_coef))
cases["CORE070-POSTFIT-COEF-NAMED-NATIVE"] = Dict(
    "pass" => coefficient_delta <= tol["coefficient_delta"], "delta" => coefficient_delta,
)

cases["CORE070-POSTFIT-CONFINT-METHODS-WALD-NATIVE"] = if j_ci_ok && r_ci_ok &&
        length(j_ci_lower) == length(r_ci_lower) == length(j_coef)
    d = maximum(abs.(vcat(j_ci_lower, j_ci_upper) .- vcat(r_ci_lower, r_ci_upper)))
    Dict("pass" => d <= tol["wald_ci_bound_delta"], "delta" => d)
else
    Dict("pass" => false, "julia_ci_ok" => j_ci_ok, "r_ci_ok" => r_ci_ok, "r_ci_error" => r_ci_error,
         "julia_len" => length(j_ci_lower), "r_len" => length(r_ci_lower))
end

response_delta = maximum(abs.(j_response .- r_response))
cases["CORE070-POSTFIT-FITTED-DEFAULT-NATIVE"] = Dict(
    "pass" => response_delta <= tol["link_response_residual_delta"], "delta" => response_delta,
)

cases["CORE070-POSTFIT-LOGLIK-DF-NATIVE"] = Dict("pass" => j_dof == r_df, "julia" => j_dof, "r" => r_df)

cases["CORE070-POSTFIT-LOGLIK-NOBS-NATIVE"] = Dict(
    "pass" => j_nobs == r_loglik_nobs_attr, "julia" => j_nobs, "r" => r_loglik_nobs_attr,
)

loglik_delta = abs(j_loglik - r_loglik)
cases["CORE070-POSTFIT-LOGLIK-VALUE-NATIVE"] = Dict(
    "pass" => loglik_delta <= tol["loglik_delta"], "delta" => loglik_delta,
)

cases["CORE070-POSTFIT-NOBS-COUNT-NATIVE"] = Dict("pass" => j_nobs == r_nobs, "julia" => j_nobs, "r" => r_nobs)
cases["CORE070-POSTFIT-NOBS-FALLBACK-NATIVE"] = Dict("pass" => j_nobs == r_nobs, "julia" => j_nobs, "r" => r_nobs)

link_delta = maximum(abs.(j_link .- r_link))
cases["CORE070-POSTFIT-PREDICT-DEFAULT-NATIVE"] = Dict(
    "pass" => link_delta <= tol["link_response_residual_delta"] &&
              r_predict_type_default == "link",
    "delta" => link_delta, "r_default" => r_predict_type_default, "julia_default" => "response",
)

cases["CORE070-POSTFIT-RE-FORM-FULL-NATIVE"] = Dict(
    "pass" => link_delta <= tol["link_response_residual_delta"], "delta" => link_delta,
)

residual_delta = maximum(abs.(j_residual .- r_residual))
cases["CORE070-POSTFIT-RESIDUAL-CONDITIONAL-NATIVE"] = Dict(
    "pass" => residual_delta <= tol["link_response_residual_delta"], "delta" => residual_delta,
)
cases["CORE070-POSTFIT-RESIDUAL-SCALES-NATIVE"] = Dict(
    "pass" => residual_delta <= tol["link_response_residual_delta"] &&
              r_residual_scale_default == "normal",
    "delta" => residual_delta,
)
cases["CORE070-POSTFIT-RESIDUAL-TYPES-NATIVE"] = Dict(
    "pass" => residual_delta <= tol["link_response_residual_delta"] &&
              r_residual_type_default == "randomized_quantile",
    "delta" => residual_delta,
)

cases["CORE070-POSTFIT-SIMULATE-DEFAULT-NATIVE"] = Dict(
    "pass" => r_simulate_condition_on_re_default == true &&
              !julia_simulate_has_condition_on_re_kwarg,
    "r_condition_on_re_default" => r_simulate_condition_on_re_default,
    "julia_has_condition_on_re_kwarg" => julia_simulate_has_condition_on_re_kwarg,
)

# ---------------------------------------------------------------------------
# Negative controls: deliberately-wrong comparisons that MUST fail.
# ---------------------------------------------------------------------------
neg_coef_delta = maximum(abs.(j_coef .- (r_coef .+ 1.0)))
negative_controls = Dict(
    "NEG-COEF-SHIFTED" => Dict("behaved" => neg_coef_delta > tol["coefficient_delta"], "delta" => neg_coef_delta),
    "NEG-LOGLIK-SHIFTED" => Dict(
        "behaved" => abs(j_loglik - (r_loglik + 1.0)) > tol["loglik_delta"],
        "delta" => abs(j_loglik - (r_loglik + 1.0)),
    ),
)

all_positive_pass = all(v["pass"] for v in values(cases))
negatives_behaved = all(v["behaved"] for v in values(negative_controls))
overall_ok = all_positive_pass && negatives_behaved

report = Dict(
    "status" => overall_ok ? "PASS" : "FAIL",
    "area" => "postfit-policy",
    "scope" => "CORE070_POSTFIT_POLICY_BATCH",
    "case_count" => length(cases),
    "negative_control_count" => length(negative_controls),
    "all_positive_pass" => all_positive_pass,
    "negative_controls_behaved_as_expected" => negatives_behaved,
    "all_checks" => overall_ok,
    "cases" => cases,
    "negative_controls" => negative_controls,
    "julia_version" => string(VERSION),
)
open(out_path, "w") do io
    print(io, to_json(report))
end
println("CORE070_POSTFIT_POLICY_BATCH_RESULT ",
        overall_ok ? "PASS" : "FAIL",
        " positive=", all_positive_pass, " negatives=", negatives_behaved)
overall_ok || exit(1)
